rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,238 @@
+marker creation requests.
+
+We did a performance test for async conflict checker using local FS and two concurrency writers based on 1500+ inflight
+markers located in MARKER0~MARKER19. It takes 179ms to finish checking conflict.
+
+Here we need to pay attention to the following details:
+
+Firstly, There may be a concurrency issue between reading and writing the same MARKER file causing EOFException. In
+order to simplify the model, we return empty result of markers when catching any exception during reading specific
+MARKER and let the subsequent checker to make up for it. For example, writer1 reads writer2's MARKER0 file, but at this
+moment writer2 is overwriting MARKER0 file. Finally, we get empty result after writer1 reading writer2's MARKER0. In
+extreme cases, the delay of early conflict detection will also occur. Fortunately, we can adjust the frequency of the
+schedule according to the actual situation.
+
+Secondly, the marker checker runnable only looks at the instants smaller than the current writer for conflict detection.
+This is to avoid two writers discovering conflicts with each other at the same time and failing together. Such conflict
+detection logic based on the instant time is consistent with the existing pre-commit conflict detection.
+
+### [2] Check Commit Conflict: Why we still need to check commit conflict here?
+
+As we know, when a writer completes writing and commits, the corresponding markers are deleted immediately. Other
+writers miss the conflict detection if solely based on the markers.
+
+Let's take Figure 4 as an example
+
+![](figure4.png)
+
+Writer1 starts writing data at time t1 and finishes at time t3. Writer2 starts writing at time t2, and updates a file
+already updated by writer1 at time t4. Since all markers of writer1 have been deleted at time t4, such conflict cannot
+be found in the stage of marker conflict detection until starting to commit. In order to avoid such delay of early
+conflict detection, it is necessary to add the steps of checking commit conflict during the detection.
+
+```
+1. Get and cache all instants before init write handler as set1.
+2. Create and reload a new active timeline after finishing marker conflict checking as set2.
+3. Get the diff instants between set1 and set2 as set3.
+4. Read all the changed fileIds contained in set3 as fileId set1
+5. Check if current writing fileID set2 is overlapped with fileId set1
+```
+
+### [3] Try to resolve conflict
+
+For now, the default behavior is that the write handlers with conflict throw an `HoodieEarlyConflictDetectionException`
+which are running on executors. These tasks then fail and the Hudi write transaction retries.
+
+As for HoodieDeltaStreamer, when we detect marker conflicts, corresponding writing task fails and retries. If the retry
+reaches a certain number of times we set, the current stage will fail. At this time, this behavior is consistent with
```

Comment:
We should think about how to abort the Hudi write transaction as a whole if one executor fails to create a marker due to early conflict detection, instead of relying on Spark's failure retries.

Reply:
Sure, Ethan! For now when hudi detected a conflict, it will throw a `HoodieEarlyConflictDetectionException` on executor ==> current task failed directly 

If we want to control this failed ourselves, there is the way i am thinking:

we implement a SparkListener for SparkListenerTaskEnd and register to spark context. If there is any task failed due to `HoodieEarlyConflictDetectionException` then failed current write transaction

Reply:
We can do sth like this and take it as a follow-up.  Have you measured how much additional time this can save, i.e., how much time do retries take?

Reply:
If there is a conflict and users use `AsyncTimelineMarkerEarlyConflictDetectionStrategy`, it will take about 150 ms to detect this conflict and fail current task. 

spark.task.maxFailures with default value 4. It may takes 4* 150 = 600ms to fail current write transaction

Reply:
Sg.  Let's follow up with proper benchmarking on this.  We should show how this contributes to the overall write latency of a transaction.  Spark retries the whole job/stage which contains multiple tasks.   Considering Spark's overhead, it can take some time for all tasks to fail and retry.